### PR TITLE
Initial impl of attestation response caching

### PR DESCRIPTION
**Pull request by waynz0r**
## Type of Change

- [x] New Feature

### CODE REVIEW

File: cert_tools.c
Comment (baluchicken) [inline cert_tools.c:1]: Thanks for fixing this.

File: socket.c [hunk 1]
```diff
@@ -765,13 +768,96 @@ static opa_socket_context socket_eval(const char *input)
...
+
+	snprintf(buff, sizeof(buff), "destination:ip:%s", destination_ip);
+	json_object_set_boolean(selectors, buff, true);
+	snprintf(buff, sizeof(buff), "destination:port:%d", destination_port);
+	json_object_set_boolean(selectors, buff, true);
+
+	char *serialized_string = json_serialize_to_string(json);
+
+	command_answer *answer = kzalloc(sizeof(struct command_answer), GFP_KERNEL);
+	answer->answer = strdup(serialized_string);
```
Comment (bonifaido) [inline socket.c:843]: No need to dup the string `serialized_string` here, just pass it as the answer, and we can remove the `json_free_serialized_string` call from here as well (which is just a `kfree`).

**Review by bonifaido — APPROVED** [review]

File: attest.c [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+    attest_response *response = kzalloc(sizeof(attest_response), GFP_KERNEL);
+
+    kref_init(&response->kref);
+
+    return response;
+}
+
+static void attest_response_free(attest_response *response)
+{
+    kfree(response->error);
```
Comment (baluchicken) [inline attest.c:81]: I would add a Null check here for response.

File: attest.c [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+    snprintf(key, keylen + 1, "%s %d %d %u %s", ctx->cgroup_path, ctx->uid.val, ctx->gid.val, ctx->namespace_ids.mnt, ctx->command_path);
+
+    return key;
+}
+
+static void attest_response_cache_set_locked(char *key, attest_response *response)
+{
+    if (!key)
+    {
+        pr_err("attest response cache: provided key is null");
```
Comment (baluchicken) [inline attest.c:128]:
```suggestion
        pr_err("nasp: attest response cache: provided key is null");
```

File: attest.c [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+        return;
+    }
+
+    housekeep_cache_locked();
+
+    attest_response_cache_entry *new_entry = kzalloc(sizeof(attest_response_cache_entry), GFP_KERNEL);
+    if (!new_entry)
+    {
+        pr_err("attest response cache: memory allocation error");
```
Comment (baluchicken) [inline attest.c:137]:
```suggestion
        pr_err("nasp: attest response cache: memory allocation error");
```

**Review by baluchicken — APPROVED** [review]

**Pull request merged**